### PR TITLE
Implement ADAPTIVE Retry Mode

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
This PR introduces support for `ADAPTIVE` retry mode.

ADAPTIVE retry mode builds on top of STANDARD retry mode and includes rate
limiting of requests when the client is throttled.

## Description
Implement ADAPTIVE retry mode. ADAPTIVE mode builds on STANDARD retry mode and adds rate limiting of requests when the client is throttled.

Adaptive retry mode dynamically limits the rate of AWS requests to maximize success rate. This may be at the expense of request latency. Adaptive retry mode is not recommended when predictable latency is important.

## Testing
 - New unit tests
 - ~~Running~~ integ tests
 - ~~Running~~ stability tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java [hunk 1]
```diff
@@ -97,7 +115,19 @@ public void maybeAttemptExecute(CompletableFuture<Response<OutputT>> future) {
             Duration backoffDelay = retryableStageHelper.getBackoffDelay();
             if (!backoffDelay.isZero()) {
                 retryableStageHelper.logBackingOff(backoffDelay);
-                scheduledExecutor.schedule(() -> attemptExecute(future), backoffDelay.toMillis(), MILLISECONDS);
+            }
+
+            OptionalDouble tokenAcquireTimeSeconds = retryableStageHelper.getSendTokenNonBlocking();
+            if (!tokenAcquireTimeSeconds.isPresent()) {
+                future.completeExceptionally(SdkClientException.create("unable to acquire send token"));
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java:122]: Can we add more context in the error message to help users mitigate this issue?
Something like
https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyRequestExecutor.java#L334=L350

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java [hunk 2]
```diff
@@ -308,6 +319,20 @@ public int hashCode() {
          * @see #numRetries(Integer)
          */
         Integer numRetries();
+
+        /**
+         * Whether the client should immediately fail the request when not enough capacity is immediately available from the
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java:324]: Can we add a note here to clarify that this only applies to adaptive mode?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java [hunk 3]
```diff
@@ -62,8 +62,8 @@ private RetryPolicy(BuilderImpl builder) {
         this.numRetries = builder.numRetries;
         this.retryCondition = builder.retryCondition;
         this.retryCapacityCondition = builder.retryCapacityCondition;
-
         this.aggregateRetryCondition = generateAggregateRetryCondition();
+        this.fastFailRateLimiting = builder.isFastFailRateLimiting();
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java:66]: Should we throw validation exception if this is provided when retry mode is not adaptive?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java [hunk 4]
```diff
@@ -97,7 +115,19 @@ public void maybeAttemptExecute(CompletableFuture<Response<OutputT>> future) {
             Duration backoffDelay = retryableStageHelper.getBackoffDelay();
             if (!backoffDelay.isZero()) {
                 retryableStageHelper.logBackingOff(backoffDelay);
```
[the remark this answers is not in the document]
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java:117]: I moved it after the `acquire()` call

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java (hunk 3 above)
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryPolicy.java:67]: Done. Switched to the nullable `Boolean` as well

**Review by zoewangg — APPROVED** [review]

**Pull request merged**